### PR TITLE
Micro-optimizes _SendSignal a bit

### PR DESCRIPTION
**Pull request by Watermelon914**
## About The Pull Request
Instead of iterating over an assoc list and doing a list access, which has a complexity of O(nlog(n)), it is better to just store a 2 tuple and access that to get a complexity of O(n)
Check code to see what I mean.

## Why It's Good For The Game
image
This is a very hot proc and it's worth micro-optimizing where we can.

The speed increase in doing the following code can be seen here:
image
Higher is better.

The code that was benchmarked:
```dm
var/list/target = list()

/proc/testa()
    var/list/queued_calls = list()
    for(var/i in 1 to length(target))
        var/data = target[i]
        queued_calls.Add(data, 1)
    for(var/i in 1 to (length(queued_calls) / 2))
        var/a = queued_calls[i*2-1]
        var/b = queued_calls[i*2]

/proc/testb()
    var/list/queued_calls = list()
    for(var/data in target)
        queued_calls[data] = 1
    for(var/data in queued_calls)
        var/a = data
        var/b = queued_calls[data]

MAIN
	for(var/i in 1 to 100)
		target.Add("[i]")
    BEGIN_BENCH(2)
        BENCH_PHASE("New code", testa())
        BENCH_PHASE("Old code", testb())
    END_BENCH
```

### CODE REVIEW

File: code/datums/signals.dm [hunk 1]
```diff
@@ -120,7 +120,9 @@
 	// all the objects that are receiving the signal get the signal this final time.
 	// AKA: No you can't cancel the signal reception of another object by doing an unregister in the same signal.
 	var/list/queued_calls = list()
-	for(var/datum/listening_datum as anything in target)
-		queued_calls[listening_datum] = listening_datum._signal_procs[src][sigtype]
-	for(var/datum/listening_datum as anything in queued_calls)
-		. |= call(listening_datum, queued_calls[listening_datum])(arglist(arguments))
+	// This should be faster than the other method of iterating over a list
```
Comment (Mothblocks) [inline code/datums/signals.dm:123]: Nit: There's like 5 ways to iterate lists in BYOND, be specific

**Review by Mothblocks — CHANGES_REQUESTED** [review]
Can you split this into a 50/50 A/B test so we can check on an actual round if its faster/slower? I'm a bit skeptical and want more than just a basic BYOND bench

**Comment by Mothblocks** [comment]
As in:
```dm
/datum/proc/RegisterSignal(etc)
  var/static/x = 0
  if (++x % 2 == 0)
    RegisterSignal_ORIGINAL(etc)
  else
    RegisterSignal_NEW(etc)

/datum/proc/RegisterSignal_ORIGINAL(etc)
  // original

/datum/proc/RegisterSignal_NEW(etc)
  // new
```

**Comment by Watermelon914** [comment]
ok

**Comment by Watermelon914** [comment]
Results over around 14 rounds of testmerge:
image

The speed increase isn't super significant, it's marginally faster but the new algorithm has a better time complexity so it'll fare better as we register more signals of the same type onto the same object.

File: code/datums/signals.dm [hunk 2]
```diff
@@ -120,7 +120,9 @@
...
-	for(var/datum/listening_datum as anything in target)
-		queued_calls[listening_datum] = listening_datum._signal_procs[src][sigtype]
-	for(var/datum/listening_datum as anything in queued_calls)
-		. |= call(listening_datum, queued_calls[listening_datum])(arglist(arguments))
+	// This should be faster than doing `var/datum/listening_datum as anything in target` as it does not implicitly copy the list
+	for(var/i in 1 to length(target))
+		var/datum/listening_datum = target[i]
+		queued_calls.Add(listening_datum, listening_datum._signal_procs[src][sigtype])
+	for(var/i in 1 to (length(queued_calls) / 2))
+		. |= call(queued_calls[i*2-1], queued_calls[i*2])(arglist(arguments))
```
Comment (LemonInTheDark) [inline code/datums/signals.dm:128]:
```suggestion
	for(var/i in 1 to length(queued_calls) step 2))
		. |= call(queued_calls[i], queued_calls[i + 1])(arglist(arguments))
```

Comment (LemonInTheDark) [reply code/datums/signals.dm:128]: should be about as or faster, and s cleaner

File: code/datums/signals.dm [hunk 3]
```diff
@@ -120,7 +120,9 @@
...
 	// AKA: No you can't cancel the signal reception of another object by doing an unregister in the same signal.
 	var/list/queued_calls = list()
-	for(var/datum/listening_datum as anything in target)
-		queued_calls[listening_datum] = listening_datum._signal_procs[src][sigtype]
-	for(var/datum/listening_datum as anything in queued_calls)
-		. |= call(listening_datum, queued_calls[listening_datum])(arglist(arguments))
+	// This should be faster than doing `var/datum/listening_datum as anything in target` as it does not implicitly copy the list
+	for(var/i in 1 to length(target))
+		var/datum/listening_datum = target[i]
+		queued_calls.Add(listening_datum, listening_datum._signal_procs[src][sigtype])
```
Comment (LemonInTheDark) [inline code/datums/signals.dm:126]: you... might? get something by preallocating queued_calls's length, but that'd likely only really matter for large lists which are not a common case, would prob slow us down normally

**Review by LemonInTheDark — CHANGES_REQUESTED** [review]

File: code/datums/signals.dm [hunk 4]
```diff
@@ -120,7 +120,9 @@
...
+	for(var/i in 1 to length(queued_calls) step 2))
```
Comment (Watermelon914) [inline code/datums/signals.dm:127]:
```suggestion
	for(var/i in 1 to length(queued_calls) step 2)
```

File: code/datums/signals.dm (hunk 3 above)
Comment (LemonInTheDark) [reply code/datums/signals.dm:126]: to clarify I mean in context of list expansion, but these lists are small so it doesn't matter/would maybe hurt

**Pull request merged**